### ISSUE TITLE
Display the duration of edit/locked event

### ISSUE THREAD

**Issue by althio** [hotosm/osm-tasking-manager2#503]
Display the duration of every edit/locked event in the history/activity log of each tile/task.
With fuzzy duration preferably (timeago-style).

As per @pgiraud in https://github.com/hotosm/osm-tasking-manager2/issues/452#issuecomment-67728592

> I agree that having the duration is more useful.
> Let's create a new issue for that.

Original idea from @pyrog in https://github.com/hotosm/osm-tasking-manager2/issues/452

With support from @althio @ethan-nelson @bgirardot @pgiraud.

Related to #469.

**Comment by ethan-nelson** [comment]
I've spent some time working on this and have a partial solution. The code location and basic form would be [task.history.mako after L71](https://github.com/hotosm/osm-tasking-manager2/blob/master/osmtm/templates/task.history.mako#L71) with:
``` 
% if (section != 'project') and isinstance(step, TaskLock) and not step.lock:
      <em>${_('for duration')}: ${str((step.date - history[index-1].date))[:-7]}</em>  
% endif
```
The issue is trying to reference the previous *lock* event. Assuming that the history had each unlock event directly after the lock event, this would work by referencing the history item directly before it. So, I'm looking for a more elegant way to find the previous lock event without making something too computationally expensive (e.g. looping backwards from that index to previous lock event). I trim the string by 7 places because the microseconds on the difference would, I believe, be more confusing than helpful.

**Comment by ethan-nelson** [comment]
Okay, update. New code:

```
    % if section != 'project' and isinstance(step, TaskLock) and not step.lock:
        % for ids in history:
            % if ids.id == step.id-1:
                <em>${_('for duration')}: ${str(step.date - ids.date)[:-7]}</em>
            % endif
        % endfor
    % endif
```
I was able to determine that the history.id attribute is similar to what I described. For this index, each lock's id is one directly below the unlock id. So, I loop over all the history events to find the id that is one below the current step's. From above, the task history is now:

![improved](https://cloud.githubusercontent.com/assets/8998918/5864701/981e7174-a249-11e4-978b-53aac6e7c4a5.JPG)

Looking for feedback. Code is available here: https://github.com/ethan-nelson/osm-tasking-manager2/commit/dd060ef18f7db7f45936e49f87d95ac1d0b34681.

**Comment by ethan-nelson** [comment]
Here's a more descriptive photo showing the complexity of step ids and events that are consecutive in the history listing: 
![solution](https://cloud.githubusercontent.com/assets/8998918/5864789/4ee50198-a24a-11e4-84a8-4c23a055508a.JPG)

**Comment by bgirardot** [comment]
I can't comment on the computational cost, but the end result is exactly what I was hoping for. I am looking forward to it. In the over all scheme of things it might not be very important, but as someone who checks a lot of work, it really helps to get an idea of how long people have worked on a task.

**Comment by althio** [comment]
Sounds very promising.
For me even the seconds are not needed. I mentioned 'fuzzy duration preferably (timeago-style)' but I don't know the work overhead.

> about a month ago for duration: 0:15:08
about a month ago for duration: 0:15 (no seconds)
about a month ago for 15 min (fuzzy)


> about a month ago for duration: 0:00:23
about a month ago for duration: 0:00 (no seconds)
about a month ago for a few seconds (or whatever fuzzy response)

**Comment by ethan-nelson** [comment]
I can set up another if statement structure to do a quasi-fuzzy style. I'll post that after I finish up some work. The timing thing I don't think is as bad of an issue because a task probably won't have tons of unlock events, and further this isn't called for the activity tab history.
Date: Thu, 22 Jan 2015 13:44:26 -0800

**Comment by althio** [comment]
Another option for display instead of timing hh:mm or fuzzy time would be
total minutes.

about a month ago for 1 minute
about a month ago for 37 minutes
about a month ago for 99 minutes

**Comment by ethan-nelson** [comment]
I like having the hours to look neater, so I'll try:
if > 60 min, display: X hour, X min.
else display: X min.

**Comment by ethan-nelson** [comment]
Better text:
![time1](https://cloud.githubusercontent.com/assets/8998918/5894180/7028febe-a4c4-11e4-9d7f-bec0a8a93465.JPG)

Also, if it's < 2 hours, it will say "1 hour, 32 minutes".

**Comment by althio** [comment]
The original idea of #452 was also to merge locked/unlocked events. It was closed via #496 but I am not certain it is completly related, see my https://github.com/hotosm/osm-tasking-manager2/issues/452#issuecomment-67750931. @ethan-nelson do you have a clear idea of what is done and not done at the moment?

If this idea of #452 [merge items] has been lost with closing/opening issues: should we open a new issue to discuss it again and eventually work on it?

Back to your present commit @ethan-nelson.
I find the result is definitively an improvement.

My preferred text would be:
> Unlocked
> [nothing on second line]

> Marked as done by FTA
> _about a month ago_

> Locked by FTA
> _about a month ago for 15 minutes_

To be discussed again if we want to hide or merge certain events.

**Comment by pgiraud** [comment]
I agree that it would be great to merge the unlocked into "locked with duration".
Sorry if I closed #452 too quickly.

**Comment by ethan-nelson** [comment]
I agree both lock/unlock entries are repetitive, especially now if we have a duration. I'll modify the new code in another branch to show this duration thing but without both lock and unlock events.
